File: SPPU-Comp-BE-2019-Course-Pattern-Laboratory-8th-Sem-Practicals-main/HPC/Lab 1/Parallel_BFSDFS.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <thread>
#include <future>
#include <atomic>
using namespace std;
class Graph {
private:
    unordered_map<int, vector<int>> adjList;
public:
    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }
    vector<int> getNeighbors(int u) {
        if (adjList.find(u) != adjList.end()) {
            return adjList[u];
        }
        return {};
    }
    void printGraph() {
        cout << "\nGraph Structure:" << endl;
        for (auto& entry : adjList) {
            cout << entry.first << " -> ";
            for (int neighbor : entry.second) {
                cout << neighbor << " ";
            }
            cout << endl;
        }
    }
};
vector<int> parallelBFS(Graph& graph, int start) {
    unordered_set<int> visited;
    queue<int> q;
    vector<int> bfsTraversal;
    atomic<bool> done(false);
    q.push(start);
    visited.insert(start);
    while (!q.empty()) {
        int levelSize = q.size();
        vector<future<void>> futures;
        for (int i = 0; i < levelSize; i++) {
            int currentNode = q.front();
            q.pop();
            bfsTraversal.push_back(currentNode);
            futures.push_back(async(launch::async, [&]() {
                for (int neighbor : graph.getNeighbors(currentNode)) {
                    if (visited.find(neighbor) == visited.end()) {
                        visited.insert(neighbor);
                        q.push(neighbor);
                    }
                }
            }));
        }
        for (auto& fut : futures) {
            fut.get();
        }
    }
    return bfsTraversal;
}
void dfsRecursive(Graph& graph, int node, unordered_set<int>& visited, vector<int>& dfsTraversal) {
    visited.insert(node);
    dfsTraversal.push_back(node);
    vector<int> neighbors = graph.getNeighbors(node);
    vector<future<void>> futures;
    for (int neighbor : neighbors) {
        if (visited.find(neighbor) == visited.end()) {
            futures.push_back(async(launch::async, [&]() {
                dfsRecursive(graph, neighbor, visited, dfsTraversal);
            }));
        }
    }
    for (auto& fut : futures) {
        fut.get();
    }
}
vector<int> parallelDFS(Graph& graph, int start) {
    unordered_set<int> visited;
    vector<int> dfsTraversal;
    dfsRecursive(graph, start, visited, dfsTraversal);   
    return dfsTraversal;
}
void printTraversalAsTree(const vector<int>& traversal, const string& traversalType) {
    cout << "\n" << traversalType << " Traversal Output: " << endl;
    for (int node : traversal) {
        cout << node << " ";
    }
    cout << endl;
}
int main() {
    Graph graph;
    int numEdges;
    cout << "Enter the number of edges you want to add to the graph:" << endl;
    cin >> numEdges;
    cout << "Enter edges as pairs of nodes (e.g., 0 1 for an edge between 0 and 1): " << endl;
    for (int i = 0; i < numEdges; i++) {
        int u, v;
        cin >> u >> v;
        graph.addEdge(u, v);
    }
    graph.printGraph();
    bool continueRunning = true;
    while (continueRunning) {
        cout << "\nChoose an option: " << endl;
        cout << "1. Parallel BFS" << endl;
        cout << "2. Parallel DFS" << endl;
        cout << "3. Exit" << endl;
        cout << "Enter your choice: ";
        int choice;
        cin >> choice;
        switch (choice) {
            case 1: {
                cout << "Enter starting node for BFS: ";
                int startBFS;
                cin >> startBFS;
                cout << "\nRunning Parallel BFS..." << endl;
                vector<int> bfsTraversal = parallelBFS(graph, startBFS);
                printTraversalAsTree(bfsTraversal, "BFS");
                break;
            }
            case 2: {
                cout << "Enter starting node for DFS: ";
                int startDFS;
                cin >> startDFS;
                cout << "\nRunning Parallel DFS..." << endl;
                vector<int> dfsTraversal = parallelDFS(graph, startDFS);
                printTraversalAsTree(dfsTraversal, "DFS");
                break;
            }
            case 3:
                continueRunning = false;
                cout << "Exited the program successfully" << endl;
                break;
            default:
                cout << "Invalid choice! Please choose a valid option." << endl;
                break;
        }
        if (continueRunning) {
            cout << "\nDo you want to continue? (yes/no): ";
            string userChoice;
            cin >> userChoice;
            if (userChoice == "no" || userChoice == "No") {
                continueRunning = false;
                cout << "Exited the program successfully" << endl;
            }
        }
    }
    return 0;
}

File: SPPU-Comp-BE-2019-Course-Pattern-Laboratory-8th-Sem-Practicals-main/HPC/Lab 1/Parallel_BFSDFS.java
import java.util.*;
import java.util.concurrent.*;
import java.util.Scanner;
public class Parallel_BFSDFS {
    static class Graph {
        private final Map<Integer, List<Integer>> adjList;
        public Graph() {
            adjList = new HashMap<>();
        }
        public void addEdge(int u, int v) {
            adjList.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
            adjList.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
        }
        public List<Integer> getNeighbors(int u) {
            return adjList.getOrDefault(u, new ArrayList<>());
        }
        public void printGraph() {
            System.out.println("\nGraph Structure:");
            for (Map.Entry<Integer, List<Integer>> entry : adjList.entrySet()) {
                System.out.print(entry.getKey() + " -> ");
                for (int neighbor : entry.getValue()) {
                    System.out.print(neighbor + " ");
                }
                System.out.println();
            }
        }
    }

    // Parallel BFS
    public static List<Integer> parallelBFS(Graph graph, int start) {
        Set<Integer> visited = Collections.synchronizedSet(new HashSet<>());
        Queue<Integer> queue = new ConcurrentLinkedQueue<>();
        List<Integer> bfsTraversal = new ArrayList<>();
        queue.add(start);
        visited.add(start);
        ExecutorService executor = Executors.newCachedThreadPool();
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Future<?>> futures = new ArrayList<>();
            for (int i = 0; i < levelSize; i++) {
                final int currentNode = queue.poll();
                bfsTraversal.add(currentNode);
                futures.add(executor.submit(() -> {
                    for (int neighbor : graph.getNeighbors(currentNode)) {
                        if (visited.add(neighbor)) {
                            queue.add(neighbor);
                        }
                    }
                }));
            }
            for (Future<?> future : futures) {
                try {
                    future.get();
                } catch (InterruptedException | ExecutionException e) {
                    e.printStackTrace();
                }
            }
        }
        executor.shutdown();
        return bfsTraversal;
    }

    // Parallel DFS
    public static List<Integer> parallelDFS(Graph graph, int start) {
        Set<Integer> visited = Collections.synchronizedSet(new HashSet<>());
        List<Integer> dfsTraversal = new ArrayList<>();
        ExecutorService executor = Executors.newCachedThreadPool();
        dfsRecursive(graph, start, visited, executor, dfsTraversal);
        executor.shutdown();
        return dfsTraversal;
    }
    private static void dfsRecursive(Graph graph, int node, Set<Integer> visited, ExecutorService executor, List<Integer> dfsTraversal) {
        visited.add(node);
        dfsTraversal.add(node);
        List<Integer> neighbors = graph.getNeighbors(node);
        List<Future<?>> futures = new ArrayList<>();
        for (int neighbor : neighbors) {
            if (!visited.contains(neighbor)) {
                futures.add(executor.submit(() -> {
                    dfsRecursive(graph, neighbor, visited, executor, dfsTraversal);
                }));
            }
        }
        for (Future<?> future : futures) {
            try {
                future.get();
            } catch (InterruptedException | ExecutionException e) {
                e.printStackTrace();
            }
        }
    }
    public static void printTraversalAsTree(List<Integer> traversal, String traversalType) {
        System.out.println("\n" + traversalType + " Traversal Output: ");
        for (int node : traversal) {
            System.out.print(node + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Graph graph = new Graph();
        System.out.println("Enter the number of edges you want to add to the graph:");
        int numEdges = scanner.nextInt();
        System.out.println("Enter edges as pairs of nodes (e.g., 0 1 for an edge between 0 and 1): ");
        for (int i = 0; i < numEdges; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            graph.addEdge(u, v);
        }
        graph.printGraph();
        boolean continueRunning = true;
        while (continueRunning) {
            System.out.println("\nChoose an option: ");
            System.out.println("1. Parallel BFS");
            System.out.println("2. Parallel DFS");
            System.out.println("3. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            switch (choice) {
                case 1:
                    System.out.print("Enter starting node for BFS: ");
                    int startBFS = scanner.nextInt();
                    System.out.println("\nRunning Parallel BFS...");
                    List<Integer> bfsTraversal = parallelBFS(graph, startBFS);
                    printTraversalAsTree(bfsTraversal, "BFS");
                    break;
                case 2:
                    System.out.print("Enter starting node for DFS: ");
                    int startDFS = scanner.nextInt();
                    System.out.println("\nRunning Parallel DFS...");
                    List<Integer> dfsTraversal = parallelDFS(graph, startDFS);
                    printTraversalAsTree(dfsTraversal, "DFS");
                    break;
                case 3:
                    continueRunning = false;
                    System.out.println("Exited the program successfully");
                    break;
                default:
                    System.out.println("Invalid choice! Please choose a valid option.");
                    break;
            }
            if (continueRunning) {
                System.out.print("\nDo you want to continue? (yes/no): ");
                String userChoice = scanner.next();
                if (userChoice.equalsIgnoreCase("no")) {
                    continueRunning = false;
                    System.out.println("Exited the program successfully");
                }
            }
        }
        scanner.close();
    }
}
